### PR TITLE
Support custom types in HiveType

### PR DESCRIPTION
**Pull request by daniel-ohayon**
When fetching metadata for Hive tables while analyzing a query, we go through the following codepath:
```
HiveMetadata.getViews -> ConnectorViewDefinition
SemiTransactionalHiveMetastore -> presto.hive.Table
...
ThriftMetastoreUtil.fromMetastoreApiTable -> presto.hive.Table
HiveMetastore.getTable -> hadoop.Table
```

Currently, the `presto.hive.Table` object can only describe types supported by open-source Hive, but not potential user-defined types.

This PR adds support for those.

### CODE REVIEW

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 1]
```diff
@@ -157,11 +157,26 @@ public SemiTransactionalHiveMetastore(
     }
 
     public synchronized Optional<Table> getTable(String databaseName, String tableName)
+    {
+        return getTable(databaseName, tableName, Optional.empty());
+    }
+
+    public synchronized Optional<Table> getTable(String databaseName, String tableName, Optional<ConnectorSession> session)
```
Comment (daniel-ohayon) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:164]: @arhimondr is this the right place to do this check to discard/enable semantic types?

This seemed like a good compromise because 
* it's above `CachingHiveMetastore` (I imagine that session param checks below the caching layer may have unpredictable results)
* but below classes that issue many calls to this method, such as `HiveMetadata`, `HiveSplitManager`, `HivePartitionManager`...

Comment (arhimondr) [reply presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:164]: From what I remember Table is not the only object containing the column type informations. Partition object also includes column type information.

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 2]
```diff
@@ -157,11 +157,26 @@ public SemiTransactionalHiveMetastore(
...
     {
         checkReadable();
         Action<TableAndMore> tableAction = tableActions.get(new SchemaTableName(databaseName, tableName));
         if (tableAction == null) {
-            return delegate.getTable(databaseName, tableName);
+            // TODO do it better – probably add semantic type instead of remove in case of leaks
+            Optional<Table> table = delegate.getTable(databaseName, tableName);
+            // TODO cannot import HiveSessionProperties class here to do this check because of circular dependency on presto-hive
+            if (table.isPresent() && session.isPresent() && !session.get().getProperty("semantic_types_enabled", Boolean.class)) {
```
Comment (arhimondr) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:172]: How strongly do we feel about keeping this settings as a session property? Do we think a configuration property is not flexible enough? Should we allow users change type semantics like that with a session property?

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/ExtendedFieldSchema.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        extends FieldSchema
+{
+    // a Presto type signature which can represent a type that is not
+    // handled by the open-source Hive TypeInfo (eg user-defined types)
+    private final TypeSignature semanticTypeSignature;
+
+    public ExtendedFieldSchema(String name, String type, String comment, TypeSignature semanticType)
+    {
+        super(name, type, comment);
+        this.semanticTypeSignature = semanticType;
```
Comment (arhimondr) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/ExtendedFieldSchema.java:29]: nit: `requireNonNull`

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveType.java [hunk 4]
```diff
@@ -115,7 +123,12 @@ public TypeInfo getTypeInfo()
 
     public TypeSignature getTypeSignature()
     {
-        return getTypeSignature(typeInfo);
+        return semanticTypeSignature.orElse(getTypeSignature(typeInfo));
+    }
+
+    public HiveType copyWithoutSemanticType()
```
Comment (arhimondr) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveType.java:129]: suggestion: I would inline this method, as it doesn't make a lot of sense as a member of this class.

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveType.java [hunk 5]
```diff
@@ -179,6 +192,13 @@ public static boolean isSupportedType(TypeInfo typeInfo)
         return false;
     }
 
+    public static HiveType valueOf(String hiveTypeName, TypeSignature semanticTypeSignature)
```
Comment (arhimondr) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveType.java:195]: Should this method be `@JsonCreator` now?

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 6]
```diff
@@ -157,11 +157,26 @@ public SemiTransactionalHiveMetastore(
     }
 
     public synchronized Optional<Table> getTable(String databaseName, String tableName)
+    {
+        return getTable(databaseName, tableName, Optional.empty());
```
Comment (arhimondr) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:161]: I would suggest not to add this override. It should either take a session or not. Otherwise it could be confusing and potentially fragile (one may expect this method to always respect the session property, when in fact it doesn't respect it in many cases)

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 7]
```diff
@@ -157,11 +157,26 @@ public SemiTransactionalHiveMetastore(
...
+
+    public synchronized Optional<Table> getTable(String databaseName, String tableName, Optional<ConnectorSession> session)
     {
         checkReadable();
         Action<TableAndMore> tableAction = tableActions.get(new SchemaTableName(databaseName, tableName));
         if (tableAction == null) {
-            return delegate.getTable(databaseName, tableName);
+            // TODO do it better – probably add semantic type instead of remove in case of leaks
+            Optional<Table> table = delegate.getTable(databaseName, tableName);
+            // TODO cannot import HiveSessionProperties class here to do this check because of circular dependency on presto-hive
```
Comment (arhimondr) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:171]: We've never used `session` before in any metastore related code. Adding this dependency is generally not desirable.

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java [hunk 8]
```diff
@@ -450,6 +452,7 @@ private static SerDeInfo getSerdeInfo(org.apache.hadoop.hive.metastore.api.Table
 
         return serdeInfo;
     }
+
```
Comment (arhimondr) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java:455]: nit: unnecessary change

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 9]
```diff
@@ -161,6 +164,12 @@ public TypeRegistry(Set<Type> types, FeaturesConfig featuresConfig)
         addParametricType(QDIGEST);
         addParametricType(TDIGEST);
 
+        addParametricType(new LongEnumParametricType("test.Hello", new LongEnumType.LongEnumMap(
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java:167]: This is a production class. I feels like we should add testing related classes only in tests

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryBatchStreamReader.java [hunk 10]
```diff
@@ -81,7 +82,7 @@ public LongDictionaryBatchStreamReader(Type type, StreamDescriptor streamDescrip
             throws OrcCorruptionException
     {
         requireNonNull(type, "type is null");
-        verifyStreamType(streamDescriptor, type, t -> t instanceof BigintType || t instanceof IntegerType || t instanceof SmallintType || t instanceof DateType);
+        verifyStreamType(streamDescriptor, type, t -> t instanceof BigintType || t instanceof IntegerType || t instanceof SmallintType || t instanceof DateType || t instanceof LongEnumType);
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryBatchStreamReader.java:85]: How confident are we that the `LongDictionaryBatchStreamReader` and the `LongDirectBatchStreamReader` are the only classes that require this modification?

**Review by arhimondr — COMMENTED** [review]
some comments

**Comment by arhimondr** [comment]
@highker @wenleix Could you please also have a look?

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 2 above)
Comment (daniel-ohayon) [reply presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:172]: The reason I feel pretty strongly about this is to facilitate deployment – a per-query opt-in limits the blast radius of these kind of changes.
Consider the following scenario: somebody changes the type of a VARCHAR column to an enum on an existing table – now all the queries that query this table can potentially break as a result. Whereas with a per-query flag we have more granular control of how we roll this out.
Wdyt?

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryBatchStreamReader.java (hunk 10 above)
Comment (daniel-ohayon) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryBatchStreamReader.java:85]: I grepped for this pattern but I am not confident. What tests could I run to be more confident?

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 2 above)
Comment (wenleix) [reply presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:172]: Can we just use `HiveSessionProperties#isSemanticTypesEnabled`? (similar to how other session properties get fetched)? -- we avoid hard code session property string like `semantic_types_enabled`

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 11]
```diff
@@ -157,11 +157,26 @@ public SemiTransactionalHiveMetastore(
...
+            if (table.isPresent() && session.isPresent() && !session.get().getProperty("semantic_types_enabled", Boolean.class)) {
+                return Optional.of(Table.builder(table.get())
+                        .setDataColumns(table.get().getDataColumns().stream()
+                                .map(c -> new Column(c.getName(), c.getType().copyWithoutSemanticType(), c.getComment()))
+                                .collect(toList()))
```
Comment (wenleix) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:176]: nit: `toImmutableList()`

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 6 above)
Comment (wenleix) [reply presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:161]: Is the reason there are `getTable` call where there is no `session` in the context? -- otherwise i would recommend to always have `session` to keep consistent behavior.

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 1 above)
Comment (wenleix) [reply presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:164]: Alternatively , can we just make the third parameter a boolean and states `discardSemanticType` instead of passing the whole `session` and only use one boolean there?

**Review by wenleix — COMMENTED** [review]
In general looks good to me. I would pass to @rongrong  for approval since she had most of the context?
